subsys/bluetooth/services/ras/rrsp/ras_rrsp.c

```diff
@@ -8,21 +8,684 @@
 #include <zephyr/bluetooth/uuid.h>
 #include <zephyr/bluetooth/gatt.h>
 #include <zephyr/logging/log.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/net_buf.h>
 #include <bluetooth/services/ras.h>
 
+#include "../ras_internal.h"
+
 LOG_MODULE_REGISTER(ras_rrsp, CONFIG_BT_RAS_RRSP_LOG_LEVEL);
 
+#define RRSP_WQ_STACK_SIZE 1024
+#define RRSP_WQ_PRIORITY   K_PRIO_PREEMPT(K_LOWEST_APPLICATION_THREAD_PRIO)
+K_THREAD_STACK_DEFINE(rrsp_wq_stack_area, RRSP_WQ_STACK_SIZE);
+
+NET_BUF_SIMPLE_DEFINE_STATIC(segment_buf, CONFIG_BT_L2CAP_TX_MTU);
+
+static struct bt_ras_rrsp {
+	struct bt_conn *conn;
+
+	struct ras_rd_buffer *active_buf;
+	struct k_work send_data_work;
+	struct k_work rascp_work;
+	struct k_work status_work;
+	struct k_timer rascp_timeout;
+
+	struct bt_gatt_indicate_params ondemand_ind_params;
+	struct bt_gatt_indicate_params rascp_ind_params;
+	struct bt_gatt_indicate_params rd_status_params;
+
+	uint8_t rascp_cmd_buf[RASCP_WRITE_MAX_LEN];
+	uint8_t rascp_cmd_len;
+
+	uint16_t ready_ranging_counter;
+	uint16_t overwritten_ranging_counter;
+	uint16_t segment_counter;
+	uint16_t active_buf_read_cursor;
+
+	bool streaming;
+	bool notify_ready;
+	bool notify_overwritten;
+	bool handle_rascp_timeout;
+} rrsp_pool[CONFIG_BT_RAS_RRSP_MAX_ACTIVE_CONN];
+
+static struct k_work_q rrsp_wq;
 static uint32_t ras_features; /* No optional features supported. */
 
-static ssize_t ras_features_read(struct bt_conn *conn, const struct bt_gatt_attr *attr,
-				 void *buf, uint16_t len, uint16_t offset)
+static void send_data_work_handler(struct k_work *work);
+static void rascp_work_handler(struct k_work *work);
+static void status_work_handler(struct k_work *work);
+static void rascp_timeout_handler(struct k_timer *timer);
+
+static int ondemand_rd_notify_or_indicate(struct bt_conn *conn, struct net_buf_simple *buf);
+static int rd_status_notify_or_indicate(struct bt_conn *conn, const struct bt_uuid *uuid,
+					uint16_t ranging_counter);
+
+static void rascp_send_complete_rd_rsp(struct bt_conn *conn, uint16_t ranging_counter);
+static void rascp_cmd_handle(struct bt_ras_rrsp *rrsp);
+
+static struct bt_ras_rrsp *rrsp_find(struct bt_conn *conn)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(rrsp_pool); i++) {
+		struct bt_ras_rrsp *rrsp = &rrsp_pool[i];
+
+		if (rrsp->conn == conn) {
+			return rrsp;
+		}
+	}
+
+	return NULL;
+}
+
+int bt_ras_rrsp_alloc(struct bt_conn *conn)
+{
+	struct bt_ras_rrsp *rrsp = NULL;
+
+	if (rrsp_find(conn) != NULL) {
+		return -EALREADY;
+	}
+
+	for (size_t i = 0; i < ARRAY_SIZE(rrsp_pool); i++) {
```

Comment:
Just so it's fair, I'm leaving this [comment](https://github.com/nrfconnect/sdk-nrf/pull/18737#pullrequestreview-2424175487) that I added to Sean's PR

But feel free to leave as is

Reply:
That's nifty, hadn't seen this macro!
I'll leave as-is but will keep it in mind; no in-tree users in ncs as of yet.